ironfish-cli/src/utils/s3.ts

```diff
@@ -181,4 +195,161 @@ async function uploadToBucket(
     })
 }
 
-export const S3Utils = { uploadToBucket }
+export async function downloadFromBucket(
+  s3: S3Client,
+  bucket: string,
+  keyName: string,
+  output: string,
+): Promise<void> {
+  const command = new GetObjectCommand({ Bucket: bucket, Key: keyName })
+  const response = await s3.send(command)
+  if (response.Body) {
+    const fileHandle = await fsAsync.open(output, 'w')
+    const ws = fileHandle.createWriteStream()
+
+    await pipeline(response.Body as Readable, ws)
+
+    ws.close()
+    await fileHandle.close()
+  }
+}
+
+export async function getPresignedUploadUrl(
+  s3: S3Client,
+  bucket: string,
+  keyName: string,
+  expiresInSeconds: number,
+): Promise<string> {
+  const command = new PutObjectCommand({
+    Bucket: bucket,
+    Key: keyName,
+  })
+
+  const signedUrl = await getSignedUrl(s3, command, {
+    expiresIn: expiresInSeconds,
+  })
+
+  return signedUrl
+}
+
+/**
+ * Returns an HTTPS URL to a file in S3.
+ * https://docs.aws.amazon.com/AmazonS3/latest/userguide/transfer-acceleration-getting-started.html
+ * https://docs.aws.amazon.com/AmazonS3/latest/userguide/dual-stack-endpoints.html
+ */
+export function getDownloadUrl(
+  bucket: string,
+  key: string,
+  region: { accelerated: true } | { accelerated: false; regionCode: string },
+  options?: { dualStack?: boolean },
+): string {
+  const dualStack = options?.dualStack ?? false
+
+  let regionString
+  if (region.accelerated) {
+    regionString = dualStack ? 's3-accelerate.dualstack' : 's3-accelerate'
+  } else {
+    regionString = dualStack ? `s3.dualstack.${region.regionCode}` : `s3.${region.regionCode}`
+  }
+
+  return `https://${bucket}.${regionString}.amazonaws.com/${key}`
+}
+
+export async function getBucketObjects(s3: S3Client, bucket: string): Promise<string[]> {
+  let truncated = true
+  let commandParams: ListObjectsCommandInput = { Bucket: bucket }
+  const keys: string[] = []
+
+  while (truncated) {
+    const command = new ListObjectsCommand(commandParams)
+    const response = await s3.send(command)
+
+    for (const obj of response.Contents || []) {
+      if (obj.Key !== undefined) {
+        keys.push(obj.Key)
+      }
+    }
+
+    truncated = response.IsTruncated || false
+    commandParams = { Bucket: bucket, Marker: response.Contents?.slice(-1)[0]?.Key }
+  }
+
+  return keys
+}
+
+export async function copyBucketObject(
```

Comment:
Should we get rid of this? It's not used anywhere now that we're re-uploading

Reply:
yea good idea